flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/netty/NettyConnectionReaderTest.java

```diff
@@ -0,0 +1,203 @@
+        CompletableFuture<Tuple2<Integer, Boolean>> availableAndPriorityConsumer =
+                new CompletableFuture<>();
+        CompletableFuture<Tuple2<Integer, Integer>> prioritySequenceNumberConsumer =
+                new CompletableFuture<>();
+        CompletableFuture<Integer> requiredSegmentIdFuture = new CompletableFuture<>();
+        Supplier<InputChannel> inputChannelSupplier =
+                createInputChannelSupplier(bufferNumber, true, requiredSegmentIdFuture);
+        NettyConnectionReader reader =
+                createNettyConnectionReader(
+                        inputChannelSupplier,
+                        createAvailableAndPriorityConsumer(availableAndPriorityConsumer),
+                        createPrioritySequenceNumberConsumer(prioritySequenceNumberConsumer));
+        Optional<Buffer> buffer = reader.readBuffer(0);
+        assertThat(buffer.isPresent()).isTrue();
+        assertThat(buffer.get().isBuffer()).isFalse();
+        assertThat(requiredSegmentIdFuture.isDone()).isFalse();
+        Tuple2<Integer, Boolean> result1 = availableAndPriorityConsumer.get();
+        assertThat(result1.f0).isEqualTo(INPUT_CHANNEL_INDEX);
+        assertThat(result1.f1).isEqualTo(true);
+        Tuple2<Integer, Integer> result2 = prioritySequenceNumberConsumer.get();
+        assertThat(result2.f0).isEqualTo(INPUT_CHANNEL_INDEX);
+        assertThat(result2.f1).isEqualTo(0);
+    }
+
+    @Test
+    void testReadEmptyBuffer() {
+        int bufferNumber = 0;
+        CompletableFuture<Tuple2<Integer, Boolean>> availableAndPriorityConsumer =
+                new CompletableFuture<>();
+        CompletableFuture<Tuple2<Integer, Integer>> prioritySequenceNumberConsumer =
+                new CompletableFuture<>();
+        CompletableFuture<Integer> requiredSegmentIdFuture = new CompletableFuture<>();
+        Supplier<InputChannel> inputChannelSupplier =
+                createInputChannelSupplier(bufferNumber, false, requiredSegmentIdFuture);
+        NettyConnectionReader reader =
+                createNettyConnectionReader(
+                        inputChannelSupplier,
+                        (inputChannelIndex, priority) ->
+                                availableAndPriorityConsumer.complete(
+                                        Tuple2.of(inputChannelIndex, priority)),
+                        (inputChannelIndex, sequenceNumber) ->
+                                prioritySequenceNumberConsumer.complete(
+                                        Tuple2.of(inputChannelIndex, sequenceNumber)));
+        Optional<Buffer> buffer = reader.readBuffer(0);
+        assertThat(buffer.isPresent()).isFalse();
+        assertThat(requiredSegmentIdFuture.isDone()).isFalse();
+        assertThat(availableAndPriorityConsumer.isDone()).isFalse();
+        assertThat(prioritySequenceNumberConsumer.isDone()).isFalse();
+    }
+
+    @Test
+    void testReadDifferentSegments() throws ExecutionException, InterruptedException {
```

Comment:
There are too many duplicate codes in these tests. I'd prefer extract and reuse their common parts.